src/toxcore/dht_new/server.rs

```diff
@@ -0,0 +1,696 @@
+    along with Tox.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+
+/*!
+Functionality needed to work as a DHT node.
+This module works as a coordinator of other modules.
+*/
+
+use futures::sync::mpsc;
+use futures::future;
+use tokio_io::IoFuture;
+
+use std::io::{ErrorKind, Error};
+use std::net::SocketAddr;
+use std::collections::HashMap;
+
+use toxcore::crypto_core::*;
+use toxcore::dht_new::packet::*;
+use toxcore::dht_new::packed_node::*;
+use toxcore::dht_new::kbucket::*;
+use toxcore::dht_new::client::*;
+use toxcore::dht_new::codec::*;
+
+/// Type representing Dht UDP packets.
+//pub type DhtUdpPacket = (SocketAddr, DhtPacket);
+/// Shorthand for the transmit half of the message channel.
+type Tx = mpsc::UnboundedSender<DhtUdpPacket>;
+
+/**
+Own DHT node data.
+
+Contains:
+
+- DHT public key
+- DHT secret key
+- Close List ([`Kbucket`] with nodes close to own DHT public key)
+- ping timeout lists ([`TimeoutQueue`])
+
+# Adding node to Close List
+
+Before a [`PackedNode`] is added to the Close List, it needs to be
+checked whether:
+
+- it can be added to [`Kbucket`] \(using [`Kbucket::can_add()`])
+- [`PackedNode`] is actually online
+
+Once the first check passes node is added to the temporary list, and
+a [`NodesRequest`] request is sent to it in order to check whether it's
+online. If the node responds correctly within [`PING_TIMEOUT`], it's
+removed from temporary list and added to the Close List.
+
+[`NodesRequest`]: ../dht/struct.NodesRequest.html
+[`Kbucket`]: ../dht/struct.Kbucket.html
+[`Kbucket::can_add()`]: ../dht/struct.Kbucket.html#method.can_add
+[`PackedNode`]: ../dht/struct.PackedNode.html
+*/
+#[derive(Clone)]
+pub struct Server {
+    /// secret key
+    pub sk: SecretKey,
+    /// public key
+    pub pk: PublicKey,
+    /// Close List (contains nodes close to own DHT PK)
+    pub kbucket: Kbucket,
+    /// tx split of channel to send packet to this peer via udp socket
+    pub tx: Tx,
+    /// store client object which has send request to peer
+    pub peers_cache: HashMap<PublicKey, Client>,
+}
+
+impl Server {
+    /**
+    Create new `Server` instance.
+
+    Note: a new instance generates new DHT public and secret keys.
+
+    DHT `PublicKey` and `SecretKey` are supposed to be ephemeral.
+    */
+    pub fn new(tx: Tx, pk: PublicKey, sk: SecretKey) -> Server {
+        let kbucket = Kbucket::new(KBUCKET_BUCKETS, &pk);
+
+        debug!("Created new Server instance");
+        Server {
+            sk: sk,
+            pk: pk,
+            kbucket: kbucket,
+            tx: tx,
+            peers_cache: HashMap::new(),
+        }
+    }
+
+    /// create new client
+    pub fn create_client(&mut self, addr: &SocketAddr, pk: PublicKey) -> Client {
+        let precomputed_key = encrypt_precompute(&pk, &self.sk);
+        Client::new(precomputed_key, pk, addr.clone(), self.tx.clone())
+    }
+    /// get client from cache
+    pub fn get_client(&self, pk: &PublicKey) -> Option<Client> {
```

Comment:
`create_client` const you almost zero CPU. Please, don't precache clients for now.

Reply:
send PingRequest to peer A with ping_id = 0x01.
send PingRequest to peer B with ping_id = 0x02.
received PingRespinse from peer B -> handle_packet() need client.